### PR TITLE
util: function to extract part of an ONNX model

### PR DESCRIPTION
**Pull request by jackwish**
Sometimes, people would like to _extract_ part of an model for development, validation or other purpose.

With `onnx.util.extract`, this is doable by specifying the input and output tensor names.

### CODE REVIEW

**Comment by jackwish** [comment]
@linkerzhang @houseroad Would you mind to take a look at this PR if you have time? I am not sure where ONNX repo may accept such utils.

**Comment by gramalingam** [comment]
Looks useful to me. A quick question/comment: I think this does not handle control-flow ops (like loop/if-then-else) correctly, since the nested scopes (sub-graphs) in those ops may contain references to other tensors from the outer scope (outer graph). If that is not fixed, it may be worth documenting it in the description as a limitation.

**Comment by jackwish** [comment]
@gramalingam Thanks for your comment! Yes, I didn't try on control-flow ops, but maybe it's insignificant:
* If the _boundary_ (defined by input and output tensors) is not cutting the subgraph, control-flow ops are just like other ops.
* If the _boundary_ is cutting the subgraph, it would be a pretty tricky scenario.  But yes, this is not supported.

I will document it clear later. Thanks for your advice!

**Comment by jackwish** [comment]
The doc issue mentioned in https://github.com/onnx/onnx/pull/2994#issuecomment-697845963 has been resolved.

Would you mind to take a look at this if you have time? @gramalingam @linkerzhang @houseroad

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:30]: it would be best if you add type annotations.

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:56]: why is sorting required here?

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:37]: adapt similar naming convention as other python files, basically this should be _build_name_dict

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:113]: if input or output names are not provided then you extract the entire graph. Why not display an error instead?

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:40]: nit: rename originals to existing_tensors and new_names to something like to_be_filtered or to_be_extracted?

File: docs/PythonAPIOverview.md
Comment (askhade) [inline docs/PythonAPIOverview.md:264]: rename to extract_model?

File: onnx/utils.py
Comment (gramalingam) [reply onnx/utils.py:56]: I think it is to ensure that the generated model has inputs/outputs in the order specified by the user (via the parameters).

File: onnx/utils.py
Comment (gramalingam) [reply onnx/utils.py:113]: Interestingly, in this case it ends up being a "model optimization" that eliminates nodes that are not used (in computing any of the output values). So, the current behavior seems to have a useful usecase. But it is worth documenting.

File: onnx/utils.py
Comment (gramalingam) [inline onnx/utils.py:65]: It seems like if a graph-input X is required to compute a specified-output Y, but if X is not in the specified-inputs, we could end up producing an incorrect model. May be it would help to check for this condition?

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py]: Yes, this is to make the order of inputs/outputs of the generated model to be the same as user input. This is not necessary but having this won't make user supervised.

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py]: Yes, the _missing input-X_ won't be taken as _input of the resulted model_. But I am not planning to detect such a corner case as it seems not to be easy work. Hope that the user may avoid this, and it can be caught by `onnx.check_model()` if we really encounter some...

**Comment by jackwish** [comment]
@askhade @gramalingam Thank you for your careful review, I have updated accordingly, please help to have a look. :)

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:160]: Looks like when input_names are not provided the the extracted sub graph uses the original inputs. This should be documented.

File: onnx/utils.py
Comment (askhade) [inline onnx/utils.py:93]: To optimize I think this can be done along with _collect_reachable_nodes. This is nice to have.

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py]: All member functions have type annotations except two which are sort complex...

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py]: Remove the _fallback_ like logic - input/output names are taken as they are.

Comment (jackwish) [reply onnx/utils.py]: For the _model optimization_ like behavior, as we are extracting sub-models - which is removing _unused_ or _unneeded_ nodes and tensors - it seems to be something as expected to me :)

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py:160]: Original input/output names are used in the _first commit_, and the current ([change](https://github.com/onnx/onnx/pull/2994/commits/dfe238f82e10f2a1c294841838535747651f851a?file-filters%5B%5D=.py#diff-498e239a2d86d6bcf902378224869c79L114)) implementation simply takes them as they are - empty is empty.

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py:93]: Yes, we can do that. IMO, _two passes_ is sort clearer to clarify the logic. And, as `_collect_reachable_tensors` doesn't walk the graph but iterate nodes, I guess it won't be significant :) Anyway, I can rewrite it if we really want that.

**Comment by jackwish** [comment]
@askhade @gramalingam I accidentally clicked _Resolve conversation_ button before *Comment* button, and lost some comments in the previous update. I have added the important ones back. If any code or comment confuses you, please raise.

File: onnx/utils.py
Comment (askhade) [reply onnx/utils.py:160]: input names being empty is not valid right? Then shouldn't there be a check for this on line 167?

File: onnx/utils.py
Comment (askhade) [reply onnx/utils.py:93]: As I mentioned this is not a must have but a nice to have.

**Comment by askhade** [comment]
Apart from the 1 open comment I have regarding input names check rest of the PR looks good to me. Once you clarify fix this I will approve the PR.

Thanks for your contribution!

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py:160]: It could be surprising but an empty input name is valid actually, though not very meaningful in practice... Empty input names are cases that the extracted model is using _constant inputs_ i.e. initializers. @askhade 

The example below is a `Transpose` node with constant input, and we are extracting the model itself - as it has only one node.
```py
import onnx
from onnx import helper, TensorProto

# generate original ONNX model
M, N, K = 2, 3, 4

wdata = [i for i in range(N*K)]
weight = helper.make_tensor('weight', TensorProto.FLOAT, [N, K], wdata)

output = helper.make_tensor_value_info('output', TensorProto.FLOAT, [K, N])

transpose = helper.make_node('Transpose', ['weight'], ['output'], name='transpose', perm=(1, 0))

graph = helper.make_graph([transpose, ], 'demo', [], [output], [weight])
opset = helper.make_operatorsetid(onnx.defs.ONNX_DOMAIN, 9)
model = helper.make_model(graph, producer_name='toy-demo', ir_version=6, opset_imports=[opset])

onnx.checker.check_model(model)
onnx.save_model(model, 'example.onnx')

# extract model
onnx.utils.extract_model('example.onnx', 'extracted.onnx', list(), ['output'])
```

File: onnx/utils.py
Comment (jackwish) [reply onnx/utils.py:93]: Let's see if we can improve this in the future :)

**Comment by jackwish** [comment]
> Apart from the 1 open comment I have regarding input names check rest of the PR looks good to me. Once you clarify fix this I will approve the PR.
> 
> Thanks for your contribution!

Thank you @askhade, I have updated [the comment](https://github.com/onnx/onnx/pull/2994#discussion_r501600166) regarding your concern.

**Review by askhade — APPROVED** [review]

**Comment by askhade** [comment]
@jackwish : Please merge with master again. It is best if you give permission to code maintainers to merge with master to make this more efficient. Thanks!

**Comment by jackwish** [comment]
done @askhade

**Pull request merged**
